src/lava/proc/io/dataloader.py

```diff
@@ -0,0 +1,193 @@
+        Parameters
+        ----------
+        var : Var
+            Variable that is connected to this object's state (ref-port).
+        """
+        self.state = RefPort(var.shape)
+        self.state.connect_var(var)
+        self._post_init()
+
+
+@requires(HostCPU)
+class AbstractPyStateModel(PyLoihiProcessModel):
+    state: Union[PyRefPort, None] = None
+    gt: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, float)
+    interval: np.ndarray = LavaPyType(np.ndarray, int)
+    offset: np.ndarray = LavaPyType(np.ndarray, int)
+
+    def __init__(self, proc_params: dict) -> None:
+        super().__init__(proc_params)
+        self.sample_id = 0
+        self.gt_state = 0
+        self.dataset = self.proc_params['saved_dataset']
+
+    def gt_array(self) -> np.ndarray:
+        if np.isscalar(self.gt_state):
+            return np.array([self.gt_state])
+        return self.gt_state
+
+    def run_spk(self) -> None:
+        self.gt.send(self.gt_array())
+
+    def post_guard(self) -> None:
+        return (self.current_ts - 1) % self.interval == self.offset
+
+    def run_post_mgmt(self) -> None:
+        data, self.gt_state = self.dataset[self.sample_id]
+        self.gt_state = self.gt_array()
+        self.state.write(data)
+        self.sample_id += 1
+        if self.sample_id == len(self.dataset):
+            self.sample_id = 0
+
+
+@implements(proc=State, protocol=LoihiProtocol)
+@tag('fixed_pt')
+class PyStateModelFixed(AbstractPyStateModel):
+    state: PyRefPort = LavaPyType(PyRefPort.VEC_DENSE, np.int32)
+
+
+@implements(proc=State, protocol=LoihiProtocol)
+@tag('floating_pt')
+class PyStateModelFloat(AbstractPyStateModel):
+    state: PyRefPort = LavaPyType(PyRefPort.VEC_DENSE, float)
+
+
+# Spike Dataloader ############################################################
+class Spike(AbstractProcess):
+    """Dataloader object that sends spike for a input sample at a
+    set interval and offset (phase).
+
+    Parameters
+    ----------
+    dataset : Iterable
+        The actual dataset object. Dataset is expected to return
+        ``(spike, label/ground_truth)`` when indexed.
+    interval : int, optional
+        Interval between each data load, by default 1
+    offset : int, optional
+        Offset (phase) for each data load, by default 0
+    """
+    def __init__(
+        self,
+        dataset: Iterable,
+        interval: int = 1,
+        offset: int = 0,
+    ) -> None:
+        super().__init__()
+        self.interval = Var((1,), interval)
+        self.offset = Var((1,), offset % interval)
+        data, gt = dataset[0]
+        data_shape = data.shape[:-1] + (interval,)
+        self.data = Var(shape=data_shape, init=np.zeros(data_shape))
+        self.s_out = OutPort(shape=data.shape[:-1])  # last dimension is time
+        self.gt = OutPort(shape=(1,) if np.isscalar(gt) else gt.shape)
+        self.proc_params['saved_dataset'] = dataset
+
+
+@requires(HostCPU)
+class AbstractPySpikeModel(PyLoihiProcessModel):
```

Comment:
Same here for the ProcessModel - there is lots of code duplication with the `State` Model. Not sure if there is a good way to reduce that?

Reply:
I've refactored them. check it out